### DISCUSSION TITLE
BeforeRouteEnter() happens after mounted() after upgrade.

### DISCUSSION THREAD

**Discussion by steven-twerdochlib** [vuejs/router#2239] [Help and Questions] [answered]
Hey,
I've been upgrading my project from Vue2 so now using router v4 and vue3.
Previously, my code ran beforeRouteEnter() function and then mounted() function but now (using console.logs();), I've noticed their running the other way round with mounted() running first?
Can anyone explain this, I can't see anything in the docs about it?

**Comment by skirtles-code** [comment, marked as answer]
Are the `beforeRouteEnter` and `mounted` in the same component?

Is the logging done immediately, at the top of the hook, or do you have some asynchronous code that might be delaying the logging?

I've adjusted the example from the Vue Router docs to use those two hooks:

- [Playground](https://play.vuejs.org/#eNqFVVtv2zYU/iucOsAKFlOOW/RBU910RbBs2KVoi71Me5ClY4sNRWok5Tgw/N97eNHFTtIicCSey3cu3znUIWoKJugXHaURa1qpDDmQUkFh4F3bkiPZKNmQ2a6DWS6CgZKdARU0NPHHUW39el3RttT75mJAjfF3kQtCaKch9u7+3MhOmHj2At1mF9FlFNwxt8xA03L0X1nDrL5a3QLn0gb7IUvw6MSte+CLNkqK7ep9pxSIkDFpC1OnWRJ05HAgPzoF3XScf0AlOR4dTOJxMlHsesCPLss/mLgjRr7JoySPVr9KfCe3soEsGfXPehRrlAxu7+zpkV+W9DEzy8sp1j8M7kkS7II6S4bGYL98L59k809opHq4Zdrg4zIIPe6U5fnA5sCnLdCF7kntBYHZnnZb0IndIBlHQAod2NDkDfnXlnLwvJBZMsO0JIIJpCwdwx4vT81cH09tx9jW+L+zUApDTeuNDxaw9q1In+pPfOFi+kTx9YjjmQvYu0Ir2BQd77Gx69N+2FHVpWKtQXLOHFzYNWyk8oncCJvMhZcTrEdoyYFyuY1n3iwMLtaIlLgVsfXZp9sUqJ7xDtozFzC1rHTae2zlZ+kaN4IQYrAt1K+Fom2n67hvuMPyaP4/7gouU18rHs5WdLnqG4MLuvTCdWeMFOS65Ky8w6UYcjjfC2/4aMBPRmrsNdFgOru049DjcGAN1dM32IHgzeM4uBzenl2Ds0ka7GPsyEQ3UVlNr9NQqLK2Exgy8tO3hUnbFWavhIeh/3egHmhwe/uWzDCJvu1YZuw1o6+nSgH2qIT4QJw/ktxHPhI7vYjgx/g7lA0NHjnjxRp4uIk+OcyUZExgMWQ3b2QFnBrFGmTTR8wj0hR7DmJrahQuF8itv7N6pFNOPSfzpmjx4pICWXWl5UGh82gY2TwaubHiPKqNaXWaJJ1o77YUe5yMFtev6Eu6SCrc6YmUgm7mayXvtbsp88h11mJfo1FSwc5IyfW8aNlzIR4ZXr+mr+lVwtk6QfSEiQr2DrtvPJZpNA7Ehm3PirRjwTiov1vDcGBOii3wC3f/u5MZ1cGQaFlDefeE/Ive+5Q/KMDqdjApzhQKZ86rbz79BXt8H5TIYsfR+hvKj4DXS2dz9Ga/dKLCtCd2LtvfHGdMbD/rm70BofuibKLjIDsi33+j9DHdl/TVpIvaPHDQtNT2E4ffykucNSbC1SpVBSoly3ZPMFlWkReLxeJnd/chHBPztcRbpUnJ1aLdO3lbVBUmO0gwSi4QlqxIQX7CnwMO3hw2+K2ZWtbLaeQRfkygLMtJAilZ4N8yIETHr13KOMs=)

`beforeRouteEnter` definitely runs before `mounted`.

As mentioned in [the docs](https://router.vuejs.org/guide/advanced/navigation-guards.html#Using-the-Options-API):

> The `beforeRouteEnter` guard does **NOT** have access to `this`, because the guard is called before the navigation is confirmed, thus the new entering component has not even been created yet.

If it hasn't been created then it definitely hasn't been mounted.

**Reply by steven-twerdochlib** [reply]
Hi @skirtles-code,
Thanks for the quick response, you're right I should've put the logging at the first line of the methods, it is run correctly. It appears the console logs start to switch around at this part of code in beforeRouteEnter():
`next((vm: any) => {
            console.log(3);
            vm.randomFunction();
          });`
Is there something that has changed since Vue2/Router3 to cause this? Also what would you recommend as the best solution to this?
